wgsl/index.bs

```diff
@@ -6981,13 +6979,296 @@ See [[#statements]] and [[#function-calls]].
+
+Note: The entire graph can be destroyed at this point. The tags listed above are all that we need to remember to analyze callers of this function.
+
+### Uniformity rules for statements ### {#uniformity-statements}
+
+In order to properly deal with reconvergence we must know when there is a single way for control to flow out of a statement.
+This is achieved by following the behaviors analysis (see [[#behaviors]]).
+
+If the set of behaviors for a statement has a single element, this means that we either did not diverge within the statement, or we reconverged, and we pick the node corresponding to control flow at the start of the statement as the node corresponding to control flow at the exit of the statement, overriding the other rules below.
+
+The rules for analyzing statements take as argument both the statement itself and the node corresponding to control flow at the beginning of it (which we'll note "CF" below) and return both of the following:
+
+* A node corresponding to control flow at the exit of it
+* A set of new nodes and edges to add to the graph
+
+In the table below, `(CF1, S) -> CF2` means "run the analysis on S starting with control flow CF1, apply the required changes to the graph, and name the resulting control flow CF2".
+Similarly, `(CF1, E) -> (CF2, V)` means "run the analysis on expression E, starting with control flow CF1, apply the required changes to the graph, and name the resulting control flow node CF2 and the resulting value node V" (see next section for the analysis of expressions).
+
+We have a similar set of rules for expressions in left-value positions, that we denote by `(CF, E) => (CF, L)`. Instead of computing the node which corresponds to the uniformity of the value, it computes the node which corresponds to the uniformity of the variable we are adressing.
+
+<table class='data'>
+  <caption>Uniformity rules for statements</caption>
+  <thead>
+    <tr><th>Statement<th>New nodes<th>Recursive analyses<th>Resulting control flow node<th>New edges
+  </thead>
+  <tr><td class="nowrap">*s1*; *s2*
+      <td>
+      <td class = "nowrap">(*CF*, *s1*) -> *CF1*<br>
+          (*CF1*, *s2*) -> *CF2*
+      <td>*CF2*
+      <td>
+  <tr><td class="nowrap">if (*e*) *s1* else *s2*
+      <td>*CFinside*, *CFend*
+      <td class="nowrap">(*CF*, *e*) -> (*CF'*, *V*)<br>
+          (*CFinside*, *s1*) -> *CF1*<br>
+          (*CFinside*, *s2*) -> *CF2*
+      <td>*CFend*
+      <td class="nowrap">*CFinside* -> *V*<br>
+          *CFend* -> {*CF1*, *CF2*}
+  <tr><td class="nowrap">loop {*s1* continuing {*s2*}}
+      <td>*CF'*
+      <td class="nowrap">(*CF'*, *s1*) -> *CF1*<br>
+          (*CF1*, *s2*) -> *CF2*
+      <td>*CF'*
+      <td class="nowrap">*CF'* -> {*CF2*, *CF*}
+  <tr><td class="nowrap">loop {*s1*}
+      <td>*CF'*
+      <td class="nowrap">(*CF'*, *s1*) -> *CF1*
+      <td>*CF'*
+      <td class="nowrap">*CF'* -> {*CF1*, *CF*}
+  <tr><td class="nowrap">break;
+      <td>
+      <td>
+      <td>*CF*
```

Comment:
Is this a new node?
Similar for continue, falltrough, discard, return.

Reply:
Oh, is it given in the context from the above statement (emphasis added):

> The rules for analyzing statements take as argument both the statement itself and the node corresponding to control flow at the beginning of it **(which we’ll note "CF" below)**

So this is saying the uniformity post-condition on 'break' is the same as the uniformity pre-condition.

Reply:
Correct.